File: services/py/src/infrastructure/http/network/connector.py
from asyncio import wait_for, open_connection

from src.application.balancer import LoadBalancer
from src.core.config import Config
from src.domain.entities.address import AddressDTO
from src.infrastructure.http.network.stream import Stream


class HttpConnector:
    def __init__(self, cfg: Config, balancer: LoadBalancer) -> None:
        self._cfg = cfg
        self._balancer = balancer

    async def connect(self, addr: AddressDTO) -> Stream:
        try:
            r, w = await wait_for(
                open_connection(addr.host, addr.port),
                timeout=self._cfg.timeouts.connect_ms
            )
            return Stream(r, w)
        except (TimeoutError, ConnectionRefusedError):
            self._balancer.report_failure(addr)
            raise

File: services/go/src/internal/infrastructure/network/connector.go
package network

import (
	"context"
	"net"
	"reverse-proxy/src/internal/application/balancer"
	"reverse-proxy/src/internal/infrastructure/config"
	"time"
)

type Connector struct {
	connectTimeout time.Duration
	lb             *balancer.LoadBalancer
}

func NewConnector(cfg *config.Config, lb *balancer.LoadBalancer) *Connector {
	connect, _, _, _ := cfg.Timeouts.ToDuration()
	return &Connector{
		connectTimeout: connect,
		lb:             lb,
	}
}

func (c *Connector) Connect(up balancer.Upstream) (*Stream, error) {
	ctx, cancel := context.WithTimeout(context.Background(), c.connectTimeout)
	defer cancel()
	d := &net.Dialer{}
	conn, err := d.DialContext(ctx, "tcp", up.Addr())
	if err != nil {
		c.lb.ReportFailure(up)
		return nil, err
	}
	return NewStream(conn), nil
}
